File: abc213e.php
<?php

list($H, $W) = explode(' ', trim(fgets(STDIN)));
$sy = $sx = 0;
$gy = $H - 1;
$gx = $W - 1;
$S = [];
for($i = 0; $i < $H; $i++){
    $S[] = trim(fgets(STDIN));
}
$que = new SplQueue();
$que->enqueue([$sy, $sx]);

$inf = 10 ** 8;
$dist = array_fill(0, $H, array_fill(0, $W, $inf));
$dist[$sy][$sx] = 0;

$dy = [-1, 0, 1, 0];
$dx = [0, -1, 0, 1];

while(!$que->isEmpty()){
    list($y, $x) = $que->dequeue();

    if($y == $gy && $x == $gx){
        echo $dist[$y][$x] . PHP_EOL;
        exit;
    }

    for($i = 0; $i < 4; $i++){
        $ny = $y + $dy[$i];
        $nx = $x + $dx[$i];

        if($ny < 0 || $ny >= $H || $nx < 0 || $nx >= $W){
            continue;
        }
        if($S[$ny][$nx] == "#"){
            continue;
        }
        if($dist[$y][$x] < $dist[$ny][$nx]){
            $dist[$ny][$nx] = $dist[$y][$x];
            $que->unshift([$ny, $nx]);
        }
    }

    for($iy = -2; $iy < 3; $iy++){
        for($ix = -2; $ix < 3; $ix++){
            if(abs($iy) == abs($ix) && abs($iy) == 2){
                continue;
            }
            $ny = $y + $iy;
            $nx = $x + $ix;
            if($ny < 0 || $ny >= $H || $nx < 0 || $nx >= $W){
                continue;
            }
            if($dist[$y][$x] + 1 < $dist[$ny][$nx]){
                $dist[$ny][$nx] = $dist[$y][$x] + 1;
                $que->enqueue([$ny, $nx]);
            }
        }
    }
}

echo "-1\n";

?>

File: abc213e.py
from collections import deque

inf = 10 ** 8

H, W = map(int, input().split())
sy, sx = 0,0
gy, gx = H - 1, W - 1

S = [input() for _ in range(H)]

que = deque()
que.append([sy, sx])

dist = [[inf] * W for _ in range(H)]
dist[sy][sx] = 0

dy, dx = [-1, 0, 1, 0], [0, -1, 0, 1]

while len(que) > 0:
    y, x = que.popleft()
#    print(y, x)
    
    # 到達している時
    if y == gy and x == gx:
        print(dist[y][x])
        exit()

    # 歩行
    for i in range(4):
        ny, nx = y + dy[i], x + dx[i]

        if not(0 <= ny < H and 0 <= nx < W):
            continue
        if S[ny][nx] == "#":
            continue

        if dist[y][x] < dist[ny][nx]:
            dist[ny][nx] = dist[y][x]
            que.appendleft([ny, nx])
    
    # パンチ
    for iy in range(-2, 3):
        for ix in range(-2, 3):
            # パンチ範囲有効外
            if abs(iy) == abs(ix) == 2:
                   continue

            ny, nx = y + iy, x + ix

            if not(0 <= ny < H and 0 <= nx < W):
                continue

            if dist[y][x] + 1 < dist[ny][nx]:
                dist[ny][nx] = dist[y][x] + 1
                que.append([ny, nx])

print(-1)